hail/src/main/scala/is/hail/expr/types/physical/PCanonicalArray.scala

```diff
@@ -0,0 +1,384 @@
+package is.hail.expr.types.physical
+
+import is.hail.annotations.Region
+import is.hail.asm4s.Code
+import is.hail.annotations._
+import is.hail.asm4s._
+import is.hail.asm4s.joinpoint._
+import is.hail.expr.ir.EmitMethodBuilder
+import is.hail.utils._
+
+object PCanonicalArray  {
+  def loadLength(aoff: Long): Int =
+    Region.loadInt(aoff)
+
+  def loadLength(aoff: Code[Long]): Code[Int] =
+    Region.loadInt(aoff)
+
+  def storeLength(aoff: Long, length: Int): Unit =
+    Region.storeInt(aoff, length)
+
+  def storeLength(aoff: Code[Long], length: Code[Int]): Code[Unit] =
+    Region.storeInt(aoff, length)
+
+  def nMissingBytes(len: Code[Int]): Code[Int] = (len + 7) >>> 3
+
+  def nMissingBytes(len: Int): Long = (len + 7L) >>> 3
+}
+
+final case class PCanonicalArray(elementType: PType, required: Boolean = false) extends PArray {
+  val elementByteSize: Long = UnsafeUtils.arrayElementSize(elementType)
+
+  val contentsAlignment: Long = elementType.alignment.max(4)
+
+  override val byteSize: Long = 8
```

Comment:
The logic about PCanonicalArray values being pointers needs to go on PArray, not PCanonicalArray, since there's a ton of code elsewhere that assumes all PArrays are pointers.

Reply:
I think that's only byteSize, here.

Reply:
Override because it's implemented on PType